lib/webauthn/authenticator_data.rb

```diff
@@ -1,27 +1,44 @@
 # frozen_string_literal: true
 
+require "bindata"
 require "webauthn/authenticator_data/attested_credential_data"
+require "webauthn/error"
 
 module WebAuthn
-  class AuthenticatorData
-    RP_ID_HASH_POSITION = 0
+  class AuthenticatorDataFormatError < WebAuthn::Error; end
 
+  class AuthenticatorData < BinData::Record
     RP_ID_HASH_LENGTH = 32
     FLAGS_LENGTH = 1
     SIGN_COUNT_LENGTH = 4
 
-    SIGN_COUNT_POSITION = RP_ID_HASH_LENGTH + FLAGS_LENGTH
+    endian :big
 
-    USER_PRESENT_FLAG_POSITION = 0
-    USER_VERIFIED_FLAG_POSITION = 2
-    ATTESTED_CREDENTIAL_DATA_INCLUDED_FLAG_POSITION = 6
-    EXTENSION_DATA_INCLUDED_FLAG_POSITION = 7
+    count_bytes_remaining :data_length
+    string :rp_id_hash, length: RP_ID_HASH_LENGTH
+    struct :flags do
+      bit1 :extension_data_included
+      bit1 :attested_credential_data_included
+      bit1 :reserved_for_future_use_4
+      bit1 :reserved_for_future_use_3
+      bit1 :reserved_for_future_use_2
+      bit1 :user_verified
+      bit1 :reserved_for_future_use_1
+      bit1 :user_present
+    end
+    bit32 :sign_count
+    count_bytes_remaining :trailing_bytes_length
+    string :trailing_bytes, length: :trailing_bytes_length
 
-    def initialize(data)
-      @data = data
+    def self.deserialize(data)
+      read(data)
+    rescue EOFError
+      raise AuthenticatorDataFormatError
     end
 
-    attr_reader :data
+    def data
+      to_binary_s
+    end
 
     def valid?
       valid_length? &&
```

Comment:
Isn't `#valid_length?` already covered by the `bindata` gem parsing now?

Reply:
Actually, `bindata`'s parsing will only raise an exception if there are less than 37 bytes (fixed part of authenticator data). Right now `AuthenticatorData#valid_length?` is testing equality (unlike `AttestedCredentialData#valid?` which was testing it was greater than the fixed part)
```ruby
def valid_length?
  data_length == base_length + attested_credential_data_length + extension_data_length
end
```
Not sure if it's worth keeping it this way though, since `attested_credential_data_length` and `extension_data_length` both come from `trailing_bytes`, which comes from the parsing.

Reply:
Gotcha.

Makes sense to keep it for now then.
Thanks.